### ISSUE TITLE
Cannot use hooks because of how containers call their views

### ISSUE THREAD

**Issue by FokkeZB** [elastic/search-ui#276]
The following:

```jsx
const MyResultView = () => {
  const someHook = useHook();

  return <div>My Result</div>;
};

<Result view={MyResultView} />
```

Errors with:
https://reactjs.org/warnings/invalid-hook-call-warning.html

I think this is because of how containers call their views:

https://github.com/elastic/search-ui/blob/c68f78f9f44680f8219c8aa1e2ed5442c56936af/packages/react-search-ui/src/containers/Result.js#L43-L50

When I change that to:

```jsx
    return <View
      className={className},
      result={result}
      key={`result-${result.id.raw}`}
      onClickLink={() => this.handleClickLink(result.id.raw)}
      titleField={titleField}
      urlField={urlField}
    />;
```

Everything works fine.

**Comment by FokkeZB** [comment]
For others who run into this, you can workaround via:

```jsx
const MyResultView = () => {
  const someHook = useHook();

  return <div>My Result</div>;
};

const MyPatchedResultView = (props) => (
  <ResultView {...props} />
);

<Result view={MyPatchedResultView} />
```

**Comment by constancecchen** [comment]
For what it's worth, I've just discovered I can reproduce this even without React hooks. For some odd reason the error messaging from React is different with vs. without React hooks.

### Without errors:

<img width="546" alt="" src="https://user-images.githubusercontent.com/549407/60372793-c6337400-99b2-11e9-8d35-c7283e08261e.png">

Note that **all** of the props being passed into Layout and Results are React functional components.

### With errors:

<img width="601" alt="" src="https://user-images.githubusercontent.com/549407/60372961-52459b80-99b3-11e9-816e-49d581880282.png">

<img width="416" alt="" src="https://user-images.githubusercontent.com/549407/60372966-5671b900-99b3-11e9-932b-cf1921719719.png">

<img width="1040" alt="" src="https://user-images.githubusercontent.com/549407/60372968-57a2e600-99b3-11e9-88e0-c707b99ac573.png">

The reason I wanted to try converting `Game` to `<Game />` was to pass a prop/method from the top level App to the child Game - something I likely can't do without the ability to pass it as an actual React component.

I also do think from a consistency standpoint that it makes sense to have `view` and `renderResult` take the same React component format that Layout's `bodyContent`, `sideContent` etc. all do.

Let me know if that makes sense at all and hopefully I'm not barking up the wrong tree here!

**Comment by FokkeZB** [comment]
I think this is a different case. By using `<Game />` you're passing a Node, not a Component. There's nothing left for `<Results />` to do with it and you may as well just pass `<Game />` to `bodyContent` instead.

**Comment by JasonStoltz** [comment]
@FokkeZB @constancecchen The "view" prop is effectively just a "render prop". The pattern as the React docs show it is like this:

```
render() {
    return (
      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>

        {/*
          Instead of providing a static representation of what <Mouse> renders,
          use the `render` prop to dynamically determine what to render.
        */}
        {this.props.render(this.state)}
      </div>
    );
  }
```

Note that the render prop is being called as function `{this.props.render(this.state)}`.

I'm hesitant to make the switch from "View()" to "<View />" because I don't fully understand the functional differences between the two styles. 

I previously rendered with the latter style, but had an issue, where if someone had something like this:

```
view={() => <SomeComponent />}
```

`SomeComponent` was actually remounted on every single re-render, which was not ideal.

I think we should dig into this a big more before making a change.

**Comment by FokkeZB** [comment]
@JasonStoltz perhaps it's due to using a titlecase variable (`View`)? Here's the demo, modified to reproduce the issue: https://codesandbox.io/embed/search-ui-national-parks-example-2ph4n

Search for anything with the console open to see it error on the `ResultView` I'm passing to `Results.renderResult`.

**Comment by FokkeZB** [comment]
@JasonStoltz I narrowed it down further and it looks like this is only an issue when the container is a (class) component. When I change it to a FC it works fine: https://codesandbox.io/embed/search-ui-national-parks-example-jtn7z

Perhaps that would be an alternative? I don't see any reason why the containers couldn't be FCs, which e.g. `PagingContainer` already is. `SearchBox` and `Sorting` do use class features, but could also be updated to use React hooks instead.

**Comment by FokkeZB** [comment]
Thanks @botelastic, but this is still something we'd like to see fixed as it forces us to use workarounds.

**Comment by JasonStoltz** [comment]
I'm applying the "pinned" tag so botelastic doesn't remove this again 👍

**Comment by artemis-prime** [comment]
Jason, re "I'm hesitant to make the switch from View() to <View /> because I don't fully understand the functional differences between the two styles." ... the difference is that one is seriously non-idiomatic to React and bypasses React's internals completely.   If you simply call a function you never trigger the tracking that allows React to register and call the hooks.  In fact in certain cases that may not trigger an error, you may actually end up registering the hooks on the wrong component!   A function only becomes a component to the React core when you actually render it (as is evidenced by the workaround).  

https://kentcdodds.com/blog/dont-call-a-react-function-component

Please fix this.  It's not a minor issue for a "React Framework" ;)

(Elastic is great. Keep up the good work)

**Comment by vxgnesh** [comment]
> For others who run into this, you can workaround via:
> 
> ```js
> const MyResultView = () => {
>   const someHook = useHook();
> 
>   return <div>My Result</div>;
> };
> 
> const MyPatchedResultView = (props) => (
>   <ResultView {...props} />
> );
> 
> <Result view={MyPatchedResultView} />
> ```

Can you elaborate on the work around you wrote?  I have the same issue and i have spent a lot of time to find a solution for this bug and luckily i found yours.

**Comment by JasonStoltz** [comment]
@vxgnesh I think the example has a slight type ... `ResultView` should be `MyResultView`.

```jsx
const MyResultView = (props) => {
  // All of your prop values are here in props
  const someHook = useHook();

  return <div>My Result</div>;
};

const MyPatchedResultView = (props) => (
  <MyResultView {...props} />
);

<Result view={MyPatchedResultView} />
```

And the answer is, those props would get passed from `Result` to `MyPatchedResultView` and then down to `MyResultView`

**Comment by JasonStoltz** [comment]
Sorry folks for not fixing this issue yet 😫

**Comment by FokkeZB** [comment]
Thanks for spotting that typo @JasonStoltz - I fixed it.

**Comment by reaves-tyler** [comment]
@FokkeZB  thank you for the update and example.  Just wanted to expand on it a little bit with my simple use case.

`interface ResultsViewProps {
    className?: string;
    result?: {};
    onClickLink?: () => void;
    titleField?: string;
    urlField?: string;
}

// Custom View Definition w/ Hooks
const MyResultView = ({className, result, onClickLink, titleField, urlField}: ResultsViewProps) => {
    const history = useHistory();

    const navigate = () => {
        history.push('/some-page/some-result-id');
    };

    return (<div onClick={navigate}>{JSON.stringify(result)}</div>);
};

// Custom View Wrapper
export const ResultsView = (props: ResultsViewProps) => <MyResultView {...props} />;

// Default Results Component from Enterprise Search Reference UI export
// https://github.com/elastic/search-ui/blob/master/packages/react-search-ui/src/containers/Results.js#L47
<Results
  titleField={getConfig().titleField}
  urlField={getConfig().urlField}
  shouldTrackClickThrough={true}
  resultView={ResultsView} // Custom View
/>;`

**Comment by joemcelroy** [comment]
this change has now been released under v1.9.0 😄  LMK of any issues.
